fish-rust/src/builtins/string/shorten.rs

```diff
@@ -0,0 +1,244 @@
+use super::{
+    escape_code_length, width_without_escapes, Arguments, Direction, StringError, StringSubCommand,
+};
+use crate::builtins::shared::{io_streams_t, STATUS_CMD_ERROR, STATUS_CMD_OK};
+use crate::common::get_ellipsis_str;
+use crate::ffi::parser_t;
+use crate::wchar::{wstr, WString, L};
+use crate::wchar_ext::WExt;
+use crate::wcstringutil::fish_wcwidth_visible;
+use crate::wcstringutil::split_string;
+use crate::wgetopt::{wopt, woption, woption_argument_t::*};
+use crate::wutil::fish_wcstol;
+
+pub struct Shorten<'args> {
+    ellipsis: &'args wstr,
+    ellipsis_width: usize,
+    max: Option<usize>,
+    no_newline: bool,
+    quiet: bool,
+    shorten_from: Direction,
+}
+
+impl Default for Shorten<'_> {
+    fn default() -> Self {
+        Self {
+            ellipsis: get_ellipsis_str(),
+            ellipsis_width: width_without_escapes(get_ellipsis_str(), 0),
+            max: None,
+            no_newline: false,
+            quiet: false,
+            shorten_from: Direction::Right,
+        }
+    }
+}
+
+impl<'args> StringSubCommand<'args> for Shorten<'args> {
+    const LONG_OPTIONS: &'static [woption<'static>] = &[
+        // FIXME: documentation says it's --char
+        wopt(L!("chars"), required_argument, 'c'),
+        wopt(L!("max"), required_argument, 'm'),
+        wopt(L!("no-newline"), no_argument, 'N'),
+        wopt(L!("left"), no_argument, 'l'),
+        wopt(L!("quiet"), no_argument, 'q'),
+    ];
+    const SHORT_OPTIONS: &'static wstr = L!(":c:m:Nlq");
+
+    fn parse_opt(&mut self, optarg: Option<&'args wstr>, c: char) -> Result<(), StringError> {
+        match c {
+            'c' => {
+                self.ellipsis = optarg.unwrap();
+                self.ellipsis_width = width_without_escapes(self.ellipsis, 0);
+            }
+            'm' => self.max = Some(fish_wcstol(optarg.unwrap())?.try_into()?),
+            'N' => self.no_newline = true,
+            'l' => self.shorten_from = Direction::Left,
+            'q' => self.quiet = true,
+            _ => return Err(StringError::UnknownOption),
+        }
+        return Ok(());
+    }
+
+    fn handle(
+        &mut self,
+        _parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optind: &mut usize,
+        args: &[&wstr],
+    ) -> Option<libc::c_int> {
+        let mut min_width = usize::MAX;
+        let mut inputs = Vec::new();
+
+        let iter = Arguments::new(args, optind, streams, true);
+
+        if Some(0) == self.max {
```

Comment:
> not a fan of yoda condition

Today I learned what a Yoda condition is, thanks!

```suggestion
        if self.max == Some(0) {
```